### PR TITLE
Compute source status 

### LINKED ISSUE #48: Return Connected or Disconnected info about sources
The sources information returned by the Rest API should also contain whether a source is currently connected or disconnected. A source is considered connected if its currently streaming data. This information is essential for dashboard.

### PR DESCRIPTION
**Pull request by nivemaham**
Compute source status from `effectiveTimeFrame.endTimeStamp` of a source and `source-type-connnection-timeout` if configured for corresponding source-type. Otherwise it would use the default connection timeout.

Closes #48

### CODE REVIEW

File: radar.yml [hunk 1]
```diff
@@ -19,4 +19,12 @@ management_portal_config:
...
-    source_endpoint: api/sources/
\ No newline at end of file
+    source_endpoint: api/sources/
+
+# Timeout duration for every source-type to decide source status whether its connected or not.
+# A source-type should be defined by following the convention of producer_model as mentioned in
+# the specification in radar-schemas
+# timeout should be specified as {@code PnDTnHnMn.nS}.
+source-type-connection-timeout:
+    android_phone: 2H
```
Comment (blootsvoets) [inline radar.yml]: Shouldn't the default be `PT2H` then? Please mention the ISO 6801 spec.

File: src/integrationTest/java/org/radarcns/webapp/SourceEndPointTest.java [hunk 2]
```diff
@@ -49,6 +61,62 @@
...
+        insertMonitorStatistics(now.minus(Duration.ofMinutes(30)), now);
+
+        String requestPath = PROJECTS + '/' + PROJECT + '/' + SUBJECTS + '/' + SUBJECT + '/'
+                + SOURCES + '/' + SOURCE;
+        try (Response response = apiClient.request(requestPath, APPLICATION_JSON, Status.OK)) {
+            assertNotNull(response);
+            assertTrue(response.isSuccessful());
+
+            ObjectReader reader = RadarConverter.readerFor(Source.class);
+            Source sources = reader.readValue(response.body().byteStream());
```
Comment (blootsvoets) [inline src/integrationTest/java/org/radarcns/webapp/SourceEndPointTest.java]: Use `getJson(requestPath, Source.class, Status.OK)` here instead (from #84).

File: src/integrationTest/resources/radar.yml [hunk 3]
```diff
@@ -21,4 +21,12 @@ management_portal_config:
...
-hdfs_output_dir: /var/lib/hdfs/output
\ No newline at end of file
+hdfs_output_dir: /var/lib/hdfs/output
+
+# Timeout duration for every source-type to decide source status whether its connected or not.
+# A source-type should be defined by following the convention of producer_model as mentioned in
+# the specification in radar-schemas
+# timeout should be specified as {@code PnDTnHnMn.nS}.
+source-type-connection-timeout:
+    android_phone: 2H
```
Comment (blootsvoets) [inline src/integrationTest/resources/radar.yml]: Fix notation here as well

**Review by blootsvoets — COMMENTED** [review]
Looks good. To merge the tests, please take a look at #84 first.

File: radar.yml (hunk 1 above)
Comment (nivemaham) [reply radar.yml]: Indeed. My bad

File: src/integrationTest/java/org/radarcns/webapp/SourceEndPointTest.java (hunk 2 above)
Comment (nivemaham) [reply src/integrationTest/java/org/radarcns/webapp/SourceEndPointTest.java]: I will merge it

**Comment by nivemaham** [comment]
@blootsvoets It is merged now. Can you have a look when you have got time? Please

File: src/main/java/org/radarcns/service/SourceService.java [hunk 4]
```diff
@@ -24,6 +29,8 @@
 
     private final ManagementPortalClient managementPortalClient;
 
+    private static final Duration DEFAULT_SOURCE_CONNECTION_TIMEOUT = Duration.ofHours(6);
```
Comment (yatharthranjan) [inline src/main/java/org/radarcns/service/SourceService.java:32]: How was this decided? Was it chosen at random?

Comment (nivemaham) [reply src/main/java/org/radarcns/service/SourceService.java:32]: The default value is set to 6 HOURS and yes it is random. It is only effective if a timeout is not configured for a source-type

File: src/main/java/org/radarcns/service/SourceService.java [hunk 5]
```diff
@@ -78,16 +85,87 @@ public SourceService(SourceMonitorService sourceMonitorService, SourceCatalog so
...
+                        source.getSourceId(), sourceType);
+        return new Source()
+                .sourceId(source.getSourceId())
+                .assigned(source.getAssigned())
+                .sourceName(source.getSourceName())
+                .sourceTypeCatalogVersion(sourceType.getCatalogVersion())
+                .sourceTypeProducer(sourceType.getProducer())
+                .sourceTypeModel(sourceType.getModel())
+                .effectiveTimeFrame(effectiveTimeFrame)
+                .status(getSourceStatus(sourceType.getProducer(), sourceType.getModel(),
```
Comment (yatharthranjan) [inline src/main/java/org/radarcns/service/SourceService.java:138]: Shouldn't this be for a Builder class like (SourceBuilder). I thought Source class is a POJO

**Review by yatharthranjan — APPROVED** [review]
Looks good to me

**Review by blootsvoets — APPROVED** [review]

File: src/main/java/org/radarcns/service/SourceService.java (hunk 5 above)
Comment (nivemaham) [reply src/main/java/org/radarcns/service/SourceService.java:138]: It indeed uses the builder pattern.

**Pull request merged**